pkg/io/v2/opener.go

```diff
@@ -0,0 +1,363 @@
+		return err
+	}
+	if err := rc.b.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *opener) RangeReader(ctx context.Context, path string, offset, length int64, opts *blob.ReaderOptions) (io.ReadCloser, error) {
+	bucket, relativePath, err := o.getBucket(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	reader, err := bucket.NewRangeReader(ctx, relativePath, offset, length, opts)
+	if err != nil {
+		return nil, err
+	}
+	return readerCloser{reader, bucket}, nil
+}
+
+func (o *opener) ReadObject(ctx context.Context, path string) ([]byte, error) {
+	b, relativePath, err := o.getBucket(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	return b.ReadAll(ctx, relativePath)
+}
+
+func (o *opener) Attributes(ctx context.Context, path string) (*blob.Attributes, error) {
+	b, relativePath, err := o.getBucket(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+
+	return b.Attributes(ctx, relativePath)
+}
+
+func (o *opener) SignedURL(ctx context.Context, path string, opts *blob.SignedURLOptions) (string, error) {
+	return providers.SignedURL(ctx, o.credentials, path, opts)
+}
+
+// Lists all paths with given path. If matchFuncs are given, all must match so that the obj
+// is included in the result
+func (o *opener) ListSubPaths(ctx context.Context, path string, matchFns ...func(*blob.ListObject) bool) ([]string, error) {
+	paths := []string{}
+	if !strings.HasSuffix(path, "/") {
+		path += "/"
+	}
+	_, _, relativePath, err := providers.ParseStoragePath(path)
+	if err != nil {
+		return nil, err
+	}
+	pathPrefix := strings.TrimSuffix(path, relativePath+"/")
+
+	it, err := o.ListSubPathsIter(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	for {
+		obj, err := it.Next(ctx)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return paths, err
+		}
+		matched := true
+		for _, matchFn := range matchFns {
+			if !matchFn(obj) {
+				matched = false
+			}
+		}
+		if matched {
+			paths = append(paths, fmt.Sprintf("%s/%s", pathPrefix, obj.Key))
+		}
+	}
+	return paths, nil
+}
+
+func (o *opener) ListSubPathsIter(ctx context.Context, path string) (*blob.ListIterator, error) {
+	bucket, relativePath, err := o.getBucket(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	q := &blob.ListOptions{
+		Prefix:    relativePath + "/",
+		Delimiter: "/",
+	}
+
+	return bucket.List(q), nil
+}
+
+func (o *opener) ListSubDirs(ctx context.Context, path string) ([]string, error) {
+	return o.ListSubPaths(ctx, path, func(obj *blob.ListObject) bool {
+		return obj.IsDir
+	})
+}
+
+// Writer returns a writer that overwrites the path.
+func (o *opener) Writer(ctx context.Context, path string, opts *blob.WriterOptions) (io.WriteCloser, error) {
+	bucket, relativePath, err := o.getBucket(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	writer, err := bucket.NewWriter(ctx, relativePath, opts)
+	if err != nil {
+		return nil, err
+	}
+	return &writerCloser{writer, bucket}, nil
+}
+
+type writerCloser struct {
+	w *blob.Writer
```

Comment:
Similarly, lets embedd just the bucket and get the Writer when we need it to avoid ambiguity